golem-component-service-base/src/service/plugin_installation.rs

```diff
@@ -0,0 +1,547 @@
+// Copyright 2024-2025 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use super::component::ComponentError;
+use super::component_object_store::ComponentObjectStore;
+use super::plugin::{PluginError, PluginService};
+use crate::model::Component;
+use crate::repo::component::{record_metadata_serde, ComponentRepo};
+use crate::service::component::TransformationFailedReason;
+use golem_common::model::component::ComponentOwner;
+use golem_common::model::component_metadata::ComponentMetadata;
+use golem_common::model::plugin::{
+    AppPluginDefinition, ComponentPluginInstallationTarget, LibraryPluginDefinition,
+    PluginInstallation, PluginInstallationCreation, PluginInstallationUpdate, PluginScope,
+    PluginTypeSpecificDefinition,
+};
+use golem_common::model::{ComponentId, ComponentVersion, PluginInstallationId, RetryConfig};
+use golem_common::retries::with_retries;
+use golem_service_base::model::VersionedComponentId;
+use golem_service_base::repo::plugin_installation::PluginInstallationRecord;
+use golem_service_base::service::plugin_wasm_files::PluginWasmFilesService;
+use reqwest::multipart::{Form, Part};
+use tracing::info;
+use tracing::info_span;
+use tracing_futures::Instrument;
+use wac_graph::CompositionGraph;
+
+#[derive(Debug)]
+pub struct PluginInstallationService<Owner: ComponentOwner, Scope> {
+    component_repo: Arc<dyn ComponentRepo<Owner> + Sync + Send>,
+    object_store: Arc<dyn ComponentObjectStore + Sync + Send>,
+    plugin_service: Arc<dyn PluginService<Owner::PluginOwner, Scope> + Sync + Send>,
```

Comment:
This guy could be converted to a struct as well, I will go for it before merging